fms_extras/utils/generation.py

```diff
@@ -3,19 +3,564 @@
+    n_gen = [0] * bsize
+    n_steps = 0
+    input_ids = torch.argmax(logits, dim=2)  # b 1
+    result = [
+        torch.cat((line, input_id), dim=0)
+        for line, input_id in zip(result, list(input_ids))
+    ]
+    # block mapping max is the max blocks required for the longest prompt in the batch
+    # this is used in padding the block mapping when using cudagraphs because cudagraphs required static shapes
+    block_mapping_max = ((max(prompt_lengths) + new_tokens - 1) // 16) + 1
+    model_input_lengths = [inp_len] * (input_ids.size(0) * n_candidates)
+
+    # perform decode step
+    while min(n_gen) < new_tokens:
+        n_steps += 1
+
+        # allocate candidate sequences in cache and get the sequence ids
+        cache_data, child_sequence_ids_list = __prepare_candidate_sequences_cache_data(
+            kv_cache_manager, parent_sequence_ids, model_input_lengths, n_candidates
+        )
+
+        # Get candidate set of speculations
+        suffix_ids = speculator.generate_suffixes(
+            embeds, input_ids, threshes, n_candidates
+        )  # b k h
+        # attach speculator output to original token inputs
+        input_ids = torch.cat(
+            [input_ids.unsqueeze(1).expand(bsize, n_candidates, 1), suffix_ids], dim=-1
+        ).int()  # b k 1+h
+
+        # Apply batch flattening / tree attention if compression is good enough
+        model_input_ids, unflat_indices, flat_indices = __maybe_flatten(
+            input_ids, cache_data, flattening
+        )
+
+        # todo: This is a WIP to enable cudagraphs, currently its only for batch_size=1
+        # cudagraph requires static shapes
+        if cudagraphs:
+            __right_pad_cache_block_mapping(cache_data, block_mapping_max)
+
+        # Base model forward pass
+        output = decode_model(
+            model_input_ids,
+            cache_data=cache_data,
+            return_embeds=True,
+            use_cache=True,
+        )  # 1 n' v  OR  bk 1+h v
+
+        next_vals, embeds = __extract_decode_output(
+            output, unflat_indices, bsize, n_candidates, inp_len
+        )
+
+        next_vals_list, embeds, parent_sequence_ids = __prune_candidates(
+            input_ids, next_vals, embeds, kv_cache_manager, child_sequence_ids_list
+        )
+
+        # Update results
+        result = [torch.cat(x, dim=0) for x in zip(result, next_vals_list)]
+        input_ids = torch.stack([line[-1:] for line in next_vals_list], dim=0)  # b 1
+
+        # update number of generated tokens per sequence
+        n_gen = [
+            n_gen_i + next_vals_i.size(0)
+            for n_gen_i, next_vals_i in zip(n_gen, next_vals_list)
+        ]
+
+    # free final parent sequences from the kv-cache
+    kv_cache_manager.free_sequences(parent_sequence_ids, recursive=True)
+
+    return result, n_steps, ttft, (time.time() - start_time)
+
+
+def __right_pad_cache_block_mapping(
+    cache_data: PagedAttentionCacheData, block_mapping_max: int
+):
```

Comment:
I think this is another function that might make more sense as a method of PagedAttentionCacheData?

Reply:
moving to the PagedAttentionCacheData